# vcfSampleCompare.pl

## WHAT IS THIS:

This script sorts and (optionally) filters the rows/variants of a VCF file (containing data for 2 or more samples) based on the differences in the variant data between samples or sample groups.  Degree of "difference" is determined by either the best possible degree of separation of sample groups by genotype calls or the difference in average allelic frequency of each sample or sample group (with a gap size threshold).  The pair of samples or sample groups used to represent the difference for a variant row is the one leading to the greatest difference in consistent genotype or average allelic frequencies (i.e. observation ratios, e.g. AO/DP) of the same variant state.  If sample groups are not specified, the pair of samples leading to the greatest difference is greedily discovered and chosen to represent the variant/row.

## DETAILS

This script works with VCF files generated by freeBayes (for SNP and small nucleotide variants) or svTyper (for structural variants).  It will work with any other VCF data that includes GT or AO, RO, and DP tags in the FORMAT string.

Each row in a VCF file will be assumed to represent a variant (or variant position).  In the context of this script, there are two ways to look at differences among the samples: genotype calls and the ratio of observations of a particular variant out of the total observatons.  We'll refer to this as either "allelic frequency" or "observation ratios" throughout the documentation.

### DEFAULT SORTING BEHAVIOR

VCF rows/variants are sorted by the (maximum) degree of difference that exists between the pairs of sample groups you define.  If multiple pairs are defined, the maximum difference computed among those pairs is used in the sort.  How the degree of difference is calculated depends on whether the --genotype or --nogenotype flag is supplied.

If --genotype is supplied, sorting will be based on the degree of difference in genotype calls between the 2 sample groups.  Put simply, variants where all the genotype calls differ between sample group 1 and sample group 2 will be at the top of the results.  If the genotype calls within a group are not consistent, the rank of the row falls and it will appear lower in the results.  If all of the genotype calls between 2 sample groups are the same, the row will be at the bottom of the results.  If samples do not have genotype calls, the rank falls even lower.  The very bottom of the results will contain variant which have no genotype calls among any samples in the 2 groups.

If --nogenotype is supplied, sorting will be based on the degree of difference in allelic frequencies between the 2 samples groups.  The degree of difference between allelic frequencies will be the maximum difference in observation ratios among the samples, e.g. an 'A' in sample 1 is seen in 1 out of 10 reads that map over the variant position whereas an 'A' is seen in sample 2 in 10 out of 10 reads.  The difference in those observation ratios is 9/10 or 0.9.  The variant state (among all the observations in the 2 sample groups) with the largest difference in observation ratios between the samples in the 2 sample groups is selected to represent the row.  The difference in average ratios of each group is what is used in the sort.

### SETTING A MINIMUM GROUP SIZE

Supplying a --min-group-size affects sorting and allows you to find which samples among 2 sample groups differ (by bringing them to the top of the sorted results).  By default, all group members are used to compute maximum difference between 2 sample groups as described above.

When --min-group-size is supplied with --nogentotype, the maximum difference between the sample groups' average observation ratios is computed twice, between N members of sample group 1 and M members of sample group 2.  When comparing sample groups, the maximum difference is determined by taking the greater difference of 2 comparisons: 1. the top N observation ratios of sample group 1 versus the bottom M (inverse) observation ratios of sample group 2.  2. the bottom N observation ratios of sample group 1 versus top M observation ratios of sample group 2.  In order to avoid meaningless results, either N or M must represent a majority of their respective sample groups.  It is recommended to always set -d to the group size for 1 of the 2 groups.  --min-group-size should only be used when the groups being compared are not 2 sets of replicates.

When --min-group-size is supplied with --gentotype, the maximum difference between the sample groups' is computed in the same manner as described above for --genotype, except calls within a sample group are allowed to differ as long as there exists a subgroup of at least size --min-group-size with a consistent genotype call.

### DYNAMIC CREATION OF SAMPLE GROUPS

When pairs of sample groups are not supplied, sample groups are greedily determined on each row independently.  Up to 2 --min-group-size's can be supplied, but must not sum to more than the number of samples.  The default minimum group sizes are both 1.  Sorting is performed in the same manner, except (in the case of --nogenotype) the top N and bottom M samples compared are selected from a single list.  In the case of --genotype, the samples are ordered by genotype call abundance and assigned to the groups from either end (omitting those with no-calls).

### GROWING SAMPLE GROUPS FROM THE MINIMUM GROUP SIZE

If you have supplied a --min-group-size that is less than the number of samples defined in the group, you can allow sample groups to grow using the --grow parameter.  This allows you to identify groups of different (i.e. non-replicate) samples that share a difference with the comparison group.  Growing groups behaves differently depending on whether --genotype or --nogenotype is supplied.

If --nogenotype is supplied, grow groups is done using the --separation-gap threshold.  It uses the difference in the obervation ratio of 1 group and its inverse observation ratio in the comparison group.  For example, if you supply `--grow --nogenotype --separation-gap 0.5`, samples will be greedily* added to the 2 groups in order of their difference with the current group's observation ratio average and stops just before the difference in the averages crosses the threshold of 0.5.

If --genotype is supplied, all members of a sample group matching a genotype call in the sample group of size --min-group-size are added to the group.  If sample groups are being created dynamically and the groups have genotype calls in common, no other samples of the common genotype call will be added.

* Sample groups are seeded with members from either the bottom or top set of observation ratios.  Samples in different groups are seeded from opposite ends (top or bottom).  Samples are then traversed top-down or bottom-up and greedily added to the respective sample group in order of ascending difference from the current group average.

### FILTERING

There are 2 threshold options that can be used to filter variants that do not contain differences between the sample groups that meet the thresholds.  In --genotype mode, the threshold is --min-group-size.  In --nogenotype mode the threshold is the combination of --separation-gap and --min-group-size.

In --nogenotype mode, if the difference between the observation ratios between (all of the*) pairs of sample groups is less than the separation gap threshold, the row will not be printed.

In --genotype mode, if the (all of the*) pairs of sample groups share a common genotypoe call, the row will not be printed.

* In either case, if any pair of sample groups meets the threshold(s), the row will be printed regardless of whether or not any other pair fails the threshold(s).

### EXAMPLE

To sort based on the difference between specific samples or groups of samples, those groups can be defined on the command line using -s.  You can specify a minimum number of samples in the groups to differ.  So for example, say you have 3 wildtype (WT) replicates and you would like to see differences that all 3 WT samples have with any one of a set of 10 mutant samples.  You would do that on the command line using the sample names:

    -s "wt1 wt2 wt3" -d 3   -s "m1 m2 m3 m4 m5 m6 m7 m8 m9 m10" -d 1

The largest difference that the average observation ratio of the WT samples has with 1 of the mutant samples will be at the top of the results.

## INSTALL

cd into the vcfSampleCompare directory and run the following:

    perl Makefile.PL
    make
    sudo make install
    
## USAGE (SHORT)

vcfSampleCompare.pl -i <file*...>... [OPTIONS]

    <file*...>...       VCF input file.
    -o <sfx>            VCF outfile suffix (appended to -i).
    -u <sfx>            [STDOUT] Summary outfile suffix (appended to -i).
    -s <str ...>...     [any^] A group of sample names for difference
                        comparisons.
                        ^ See --extended usage.
    -d <int,...>...     [all] Minimum number of samples to use in a group to
                        determine difference with its partner.
    -a <flt>            [0.6] Allelic frequency difference threshold [0-1].
    --no-g              Do not use genotype calls (use allelic frequency).
    --no-f              Do not filter variant rows whose sample groups do not
                        differ (enough).
    --no-w              Do not add as many samples to sample groups as
                        possible.
    --help              Print general info and file formats.
    --extended [<cnt>]  Print detailed usage.

## USAGE (LONG)

vcfSampleCompare.pl -i <file*...>... [OPTIONS]
vcfSampleCompare.pl [OPTIONS] < input_file

    <file*...>...             VCF input file generated either by freeBayes or
    -i <file*...>...          svTyper.
    < <file>
    -i <stub> < <file>        See --help for file format.
  
    -o <sfx>                  Outfile suffix appended to file names supplied to
                              [-i].  See --help for file format.
  
    -u <sfx>                  [STDOUT] Summary outfile suffix (appended to -i).
                              This file will contain a row for each variant row in
                              the input VCF file, and include the sorting value
                              and pairs of sample groups that were identified as
                              different.
  
                              See --help for file format.
  
                              Mutually exclusive with [--outfile] (both options
                              specify an outfile name in different ways for the
                              same output).
  
    -s <str ...>...           [any^] This option must be supplied an even number
    --sample-group            of times (or once* or 0 times**).  Each pair of
      <str ...>...            samples groups, in order, is compared to determine
                              the maximum difference between the groups.  For
                              example, if you have 3 wildtype samples and 4 mutant
                              samples, you can define these 2 groups using -s 's1
                              s2 s3' -s 's4 s5 s6 s7' (where sample name 's1',
                              's2', and 's3' are the wildtype samples and 's4',
                              's5', 's6', and 's7' are mutant samples.  (All
                              sample names must match the sample names in the VCF
                              column headers row.)  The differences in variant
                              states between these groups of samples will be used
                              to sort the variants/rows of the VCF file.  See
                              --extended --help for a description of how degree of
                              difference is calculated.
  
                              ^ If no sample groups are supplied, a default pair
                              of samples that are the most different on any
                              particular row will be chosen.
                              * If only one group is defined, the second group is
                              assumed to be the remainder of the samples.
                              ** If no groups are defined, groups are dynamically
                              determined for each variant/row.  See --help
                              --extended for details.
  
    -d <int,...>...           [all] Each sample group defined by -s is accompanied
    --min-group-size          by a (minimum) number of samples in that group with
      <int,...>...            which to compute the maximum difference against its
                              partner group.  Each instance of -s should have a -d
                              value supplied.  The order of the -d values should
                              correspond to the order of the -s sample groups they
                              apply to.  The default for each group is the group
                              size, but a smaller number can bespecified.  The
                              purpose is best shown by example.  If you have 5
                              mutant samples and 3 replicate wildtype samples, you
                              may want to find variants where 1 or more mutants
                              differ from all 3 wildtype samples, thus -d for the
                              mutant group would be '1' and -d for the wildtype
                              group would be '3'.  In order to produce meaningful
                              results, one group in each pair of groups must get a
                              value that is larger than half the group size.  See
                              --help --extended for details on how this affects
                              variant sorting, sample group growing, and
                              filtering.

    -a <flt>                  [0.6] The maximum difference between average
    --separation-gap <flt>    observation ratios for a given variant state (e.g. a
                              SNP value of "A") between groups defined by -s (and
                              -d) must be at least this value in order to either
                              be retained (see --filter|--nofilter) or grown
                              (--grow|--nogrow).  This option is only used when
                              either --filter or --grow are true.  See --help
                              --extended for more details.
  
    --no-g                    Do not use the genotype call (i.e. the 'GT' value in
    --nogenotype              the FORMAT string) for sorting & filtering rows, or
                              growing sample groups.  See --nogrow and --nofilter.
                              See --help --extended for details.
  
    --no-f                    Do not filter variant rows whose sample group pairs
    --nofilter                do not meet thresholds defined by either `--genotype
                              --min-group-size <int>` or `--nogenotype --min-
                              group-size <int> --separation-gap <float>`.  When
                              --genotype is supplied, the genotype calls in all
                              samples in each minimum group must not have any
                              common genotype calls and meet the minimum size
                              requirement.  When --nogenotype is supplied, the
                              difference in average observation ratios must be
                              greater than or equal to the --separation-gap.  See
                              --help --extended for more details.
  
    --no-w                    If the -d is less than the actual sample group size,
    --nogrow                  by default, the script will keep adding samples to
                              the minimum groups (from its remaining members) as
                              long as (if --nogenotype is supplied) the difference
                              in average observation ratios between the groups is
                              greater than -a or (if --genotype is supplied) the
                              genotype call is the same as current members and
                              different from all partner group genotypes.  Note,
                              this may lower the sort order of a variant/row when
                              --nogenotype is supplied.
  
    ...

## INPUT FORMAT (-i, --vcf-file, --input-file, STDIN)

A VCF file is a plain text, tab-delimited file.  The format is generally described here: http://bit.ly/2sulKcZ and described in detail here: http://bit.ly/2gKP5bN

However, the important parts that this script relies on are:

1. The column header line (in particular - looking for the FORMAT and sample name columns).
2. The colon-delimited codes in the FORMAT column values, specifically (for SNP data produced by freeBayes and Structural Variant data produced by SVTyper) AO (the number of reads supporting the variant), RO (the number of reads supporting the reference), and DP (the number of reads that map at or over the variant position).
3. The colon-delimited values in the sample columns that correspond to the positions defined in the FORMAT column.

The file may otherwise be a standard VCF file containing header lines preceded by '##'.  Empty lines are OK and will be printed regardless of parameters supplied to this script.  Note, the --header and --no-header flags of this script do not refer to the VCF file's header, but rather the run info header of this script.


## OUTPUT FORMAT: (-o)

The output file is the same format as the input VCF files, except sorted differently and possibly filtered.

## OUTPUT FORMAT: (--outfile, -u)

Tab delimited file describing the sorting, filtering, and differing sample groups.  The columns of the file are:

* Chromosome - The chromosome on which the variant is located.
* Position - The position starting from 1 where the variant is located.
* Reference Value - The value the reference has in the variant position.
* Alternate Value(s) - The value(s) observed in the samples in the variant position.
* Best Sample Group Pair Number - The sample group pair's number (numbered from left to right, as they were supplied on the command line) for the pair that resulted in the biggest difference in variant states between the sample groups.
* Best Score - The value the sorting of the file is based on, which is the maximum "Pair Score".
* Best Average Read Depth - The average read depth of the samples in both sample groups of the pair with the best score.
* Pair Number - A colon-delimited list of numbers indicating the pair of sample groups the sort and filtering is based on.
* Pair Score - A colon-delimited list of each pair's maximum difference between sample group 1 & 2.
* Pair Average Read Depth - A colon-delimited list of each pair's average read depth of the samples in both sample groups.
* State(s) Used - The variant states used to calculate the pair score.  The state used can be 0 (the value of the variant at the variant position in the reference), or a number indicating which of the alternate observations produced the best score.  E.g. if in --nogenotype mode and the state '0' was used to compute the scores (indicating the reference state), then the average observation ratio of each group will be close to either N/N (i.e. the same as the reference) and the other group will be close to 0/N (i.e. different from the reference).  In --genotype mode, there are 2 sets of variant state values (i.e. genotype calls) separated by a semicolon.  If there are multiple genotype calls in a sample group, they will be delimited with a "+".
* Sample Group 1 Members - A colon-delimited list of each pair's comma-delimited group 1 sample names that were used to produce the sort value by determining their maximum difference.  E.g. For 2 pairs, the value might be: "s1,s2:s6,s7".
* Sample Group 1 Values - A colon-delimited list of pairs' group 1 comma-delimited values that were used to compute the corresponding "Pair Sort Value".  These may be a series of 0's and 1's indicating whether the sample's genotype is unique to group 1 or they are the observation ratios (in the form "numerator/denominator") of the selected variant state, chosen to differ most between sample groups.
* Sample Group 2 Members - A colon-delimited list of each pair's comma-delimited group 2 sample names that were used to produce the sort value by determining their maximum difference.  E.g. For 2 pairs, the value might be: "s3,s4,s5:s8,s9,s10".
* Sample Group 2 Values - A colon-delimited list of pairs' group 2 comma-delimited values that were used to compute the corresponding "Pair Sort Value".  These may be a series of 0's and 1's indicating whether the sample's genotype is unique to group 1 or they are the observation ratios (in the form "numerator/denominator") of the selected variant state, chosen to differ most between sample groups.

Example:

    #CHROM	POS	ID	REF	ALT	BEST_PAIR	BEST_SEP_SCORE	BEST_AVEDP	PAIR_NUM	PAIR_SEP_SCORE	PAIR_AVEDP	STATE(S)_USED	PAIR1_MEMBERS	PAIR1_SCORE_DATA	PAIR2_MEMBERS	PAIR2_SCORE_DATA
    Chromosome	3413302	.	T	C	1	1	6	1	1	6	1	wtsmpl	2/2	smpl1,smpl2,smpl3	0/9,0/9,0/4
    Chromosome	4904909	.	T	C	1	1	5	1	1	5	1	wtsmpl	2/2	smpl1,smpl2,smpl3	0/5,0/7,0/7
    Chromosome	747041	.	ACCAGCTGGAAAAGGCTGGCGTGAGC	AC	1	1	5	1	1	5	1	wtsmpl	0/7	smpl2,smpl3	5/5,3/3
    Chromosome	3778741	.	T	C	1	1	5	1	1	5	1	wtsmpl	0/7	smpl1	3/3
