# vcfSampleCompare.pl

## WHAT IS THIS:

This script sorts and (optionally) filters the rows/variants of a VCF file (containing data for 2 or more samples) based on the differences in the variant data between samples or sample groups.  Degree of "difference" is determined by either the best possible degree of separation of sample groups by genotype calls or the difference in average allelic frequency of each sample or sample group (with a gap size threshold).  The pair of samples or sample groups used to represent the difference for a variant row is the one leading to the greatest difference in consistent genotype or average allelic frequencies (i.e. observation ratios, e.g. AO/DP) of the same variant state.  If sample groups are not specified, the pair of samples leading to the greatest difference is greedily discovered and chosen to represent the variant/row.

## DETAILS

This script works with VCF files generated by freeBayes (for SNP and small nucleotide variants) or svTyper (for structural variants).  It will work with any other VCF data that includes GT or AO, RO, and DP tags in the FORMAT string.

Each row in a VCF file will be assumed to represent a variant (or variant position).  In the context of this script, there are two ways to look at differences among the samples: genotype calls and the ratio of observations of a particular variant out of the total observatons.  We'll refer to this as either "allelic frequency" or "observation ratios" throughout the documentation.

### SORTING

The primary sorting value is the BEST_SEP_SCORE output column (see -u).  There is a secondary sort on descending average depth (though this is only intended to move very low depth results to the bottom and has no significant meaning above any adequate depth threshold).  The values reported in the BEST_SEP_SCORE column will always be a value between 0 and 1 (or no data) and is intended to reflect the maximum "degree of difference" between sample groups.

#### SORTING METRICS

The degree of difference between sample groups can be based on genotype calls or on observation frequencies, depending on the --genotype/--nogenotype parameter.  In both cases, the values in the column reflect the maximum degree of difference between the sample groups.  The default sorting metric is --genotype.

##### GENOTYPE MODE

If run in --genotype mode, the values in the output BEST_SEP_SCORE column (see -u) reflect the degree of difference in the genotype calls among the members of each sample group.

If all the samples in both groups have genotype calls and all the genotype calls of group 1 differ from all the genotype calls of group 2 (i.e. there are no common calls), the score will be 1.  If there are any common genotype calls between the 2 groups, the score will be 0.

If not all the samples in both groups have genotype calls, but the existing calls between the 2 groups differ, the BEST_SEP_SCORE will be a value between 0 and 1.

The score of 0 can either mean that there were differing genotype calls between the sample groups or that none of the samples had a genotype call.  Despite the common score, cases of no data in all samples will be sorted to the bottom of the output and can be differentiated by the BEST_AVEDP value.

##### NO-GENOTYPE MODE (OBSERVATION FREQUENCY MODE)

If run in --nogenotype mode, the values in the output BEST_SEP_SCORE column (see -u) reflect the degree of difference in the observation frequencies (referred to as "allelic frequencies").  Each variant state is considered independently and the one resulting in the highest BEST_SEP_SCORE is the one that is used to represent the row.  Consider SNPs for example.  A SNP can have 1 of 4 values (A, T, G, or C).  One of those is the reference state (RO), which is referred to as state 0.  The other three states are 1, 2, and 3.  Each state is considered individually.  For each state, the observation frequencies are computed as (RO or AO)/DP for all of the samples.  Those frequencies are averaged for each sample group and the difference between those averages is determined.  The state resulting in the highest score is what is reported under the BEST_SEP_SCORE column (see -u).  If there is a tie between RO and one of the AO values, the AO state (1, 2, or 3) is preferred.  For example, if state 1 is the state used to calculate the BEST_SEP_SCORE, the ratio for each sample of AO(1)/DP is computed and those values for each group are averaged and the difference between them is what goes in the BEST_SEP_SCORE column.

If any sample in a group has no depth (DP=0), the score is set to 0.

#### SAMPLE GROUP CONSTRUCTION AND ITS EFFECT ON SORTING

Sample group membership is what determines the BEST_SEP_SCORE, which is the primary basis for sorting variant rows.  Thus, how groups are defined can affect the sort of the variants.  The intention is to end up with variants at the top of the file that differ the most between samples or sample groups.

The composition of the sample groups can be user-defined or dynamic and depends on the --sample-group/-s, --min-group-size/-d, and --grow/--nogrow parameters.  The default sample groups are constructed dynamically.  Since the BEST_SEP_SCORE column values are the primary basis for sorting and it is the GT, AO, RO, and DP values of the members of the sample groups that determines that value, parameters which affect the sample groupings will affect the sort.  The default behavior is to create 2 sample groups dynamically (unless --sample-group is provided).

##### USER-DEFINED SAMPLE GROUPS

If you are interested in a comparison between 2 specific sets of samples, you can define them on the command line using 2 instances of the --sample-group parameter.  For example, if you have 3 wildtype replicates (wt1, wt2, and wt3) and 3 mutant replicates (mut1, mut2, and mut3), you can define them on the command-line like this:

```
vcfSampleCompare.pl --sample-group 'wt1 wt2 wt3' --sample-group 'mut1 mut2 mut3' input.vcf
```

The values used must match the vcf file headers for those samples (appearing after the 'FORMAT' column header).  Each list is space-delimited and each group must be wrapped in quotes.

You can have multiple sample group pairs.  Each pair must occur in tandem (e.g. `-s 'pair1/group1...' -s 'pair1/group2...' -s 'pair2/group1...' -s 'pair2/group2...'`).  When given multiple pairs of sample groups to compare, only the pair resulting in the BEST_SEP_SCORE will be reported.  In the event of a tie, the first pair defined on the command line will be the one reported.

By default, all samples in each group are used to calculate the score for the pair of groups.

Alternatively, if you are interested in at least N samples of group 1 that differ by at least the --separation-gap with M members of group 2 and --grow is true, you can use the --min-group-size parameter to supply the minimum number of members for each group (let's call them N & M for groups 1 & 2 respectively).  Note, either N or M must represent at least half the members of their respective group sizes in order to avoid the introduction of noise.  Unless --nogrow is supplied, vcfSampleCompare will try to add other samples in the defined group (over the supplied N and M values) until either all members of the group are included or the difference in their observation ratio averages would drop below the --separation-gap threshold.  Each group is sorted by their observation ratios ((AO or RO)/DP).  2 attempts are made to create each group: bottom-up versus top-down (in terms of observation ratios for groups 1 & 2), and vice versa.  The first member of each group is the lowest/highest from group1/group2 respectively.  The second attempt is the reverse: highest/lowest from group1/group2.  If --grow is true (which is the default), 1 group member will be added to one of the groups, iteratively in a greedy fashion.  The group selected to grow at each iteration, using the member from the end of its list it is being constructed from, will be the one resulting in the highest separation gap.  Groups will stop growing when doing so would cause the separation gap to exceed the --separation-gap threshold or until all group members have been added.  For example:

```
vcfSampleCompare.pl --sample-group 'wt1 wt2 wt3' --min-group-size 3 --sample-group 'mut1 mut2 mut3' --min-group-size 1 --separation-gap 0.6 --grow --nogenotype input.vcf
```

In this example, the separation score will be based on all wildtype samples and at least 1 mutant sample, though up to 3 mutant samples will be included in the mutant group as long as the resulting gap between the averages of the observation ratios of the 2 groups is greater than or equal to 0.6.

If --nogrow is supplied, group sizes will reflect the --min-group-size.  Groups will not grow beyond the members defined using --sample-group.

Note, when multiple pairs of groups are submitted on the command line, each variant/row may be composed of a different pair of groups.  The selected pair is intended to represent the groupings that have greatest difference for that variant.  If there are multiple pairs of sample groups that are of interest, it is recommended to run vcfSampleCompare.pl multiple times, one for each pairing.  Supplying multiple pairs of groups is intended to be used when you do not know which samples have variants of interest.

##### DYNAMIC CREATION OF SAMPLE GROUPS

If no --sample-group(s) or --min-group-size(s) are supplied, 2 sample groups are constructed in the following manner: samples are sorted by their observation ratios and the sample at each end of the sorted list seeds 2 initial groups: one from the beginning of the list and one from the end.  Unless --nogrow is supplied (or until each group's --min-group-size's are reached), the next sample (from either end of the remaining sorted list) which results in the largest separation gap (i.e. the difference between the averages of the observation ratios in the 2 groups) is added to its end-group.  Samples continue to be added to the groups, one by one, until the difference in their average observation ratios would go below the --separation-gap threshold (or in the case of --nogrow: until each group's --min-group-size's are reached).  Note, if the --separation-gap threshold is set to 0 and --grow is true (default), all samples will end up in a group.

Up to 2 --min-group-size's can be supplied, but must not sum to more than the number of samples.  The default --min-group-size is 1 when --sample-group is not provided.

Each grouping is independent for each variant (i.e. each row in the VCF file).  Thus, when groups are dynamically created, each variant/row may be composed of different groups.  The selected groups are intended to represent the binary sample division that represents the greatest difference for that variant.  If there are more than 2 types of samples (e.g. multiple treatments and a control), dynamic creation of sample groups can miss significant differences and it is recommended to supply sample groupings in multiple runs of vcfSampleCompare.pl.

### FILTERING

There are 2 threshold options that can be used to filter variants that do not contain differences between the sample groups that meet the thresholds.  In --genotype mode, the threshold is --min-group-size.  In --nogenotype mode the threshold is the combination of --separation-gap and --min-group-size.

In --nogenotype mode, if the difference between the observation ratios between (all of the*) pairs of sample groups is less than the separation gap threshold, the row will not be printed.

In --genotype mode, if the (all of the*) pairs of sample groups share a common genotypoe call, the row will not be printed.

* In either case, if any pair of sample groups meets the threshold(s), the row will be printed regardless of whether or not any other pair fails the threshold(s).

### EXAMPLE

To sort based on the degree of difference between specific samples or groups of samples, those groups can be defined on the command line using --sample-group/-s.  You can specify a minimum number of samples in the groups to differ.  So for example, say you have 3 wildtype (WT) replicates and you would like to see differences that all 3 WT samples have with any one of a set of 10 mutant samples.  You would do that on the command line using the sample names:

```
vcfSampleCompare.pl -s "wt1 wt2 wt3" -d 3   -s "m1 m2 m3 m4 m5 m6 m7 m8 m9 m10" -d 1 input.vcf
```

The largest difference that the average observation ratio of the WT samples has with 1 of the mutant samples will be at the top of the results.

## INSTALL

cd into the vcfSampleCompare directory and run the following:

    perl Makefile.PL
    make
    sudo make install
    
## USAGE (SHORT)

vcfSampleCompare.pl -i <file*...>... [OPTIONS]

    <file*...>...       VCF input file.
    -o <sfx>            VCF outfile suffix (appended to -i).
    -u <sfx>            [STDOUT] Summary outfile suffix (appended to -i).
    -s <str ...>...     [any^] A group of sample names for difference
                        comparisons.
                        ^ See --extended usage.
    -d <int,...>...     [all] Minimum number of samples to use in a group to
                        determine difference with its partner.
    -a <flt>            [0.6] Allelic frequency difference threshold [0-1].
    --no-g              Do not use genotype calls (use allelic frequency).
    --no-f              Do not filter variant rows whose sample groups do not
                        differ (enough).
    --no-w              Do not add as many samples to sample groups as
                        possible.
    --help              Print general info and file formats.
    --extended [<cnt>]  Print detailed usage.

## USAGE (LONG)

vcfSampleCompare.pl -i <file*...>... [OPTIONS]
vcfSampleCompare.pl [OPTIONS] < input_file

    <file*...>...             VCF input file generated either by freeBayes or
    -i <file*...>...          svTyper.
    < <file>
    -i <stub> < <file>        See --help for file format.
  
    -o <sfx>                  Outfile suffix appended to file names supplied to
                              [-i].  See --help for file format.
  
    -u <sfx>                  [STDOUT] Summary outfile suffix (appended to -i).
                              This file will contain a row for each variant row in
                              the input VCF file, and include the sorting value
                              and pairs of sample groups that were identified as
                              different.
  
                              See --help for file format.
  
                              Mutually exclusive with [--outfile] (both options
                              specify an outfile name in different ways for the
                              same output).
  
    -s <str ...>...           [any^] This option must be supplied an even number
    --sample-group            of times (or once* or 0 times**).  Each pair of
      <str ...>...            samples groups, in order, is compared to determine
                              the maximum difference between the groups.  For
                              example, if you have 3 wildtype samples and 4 mutant
                              samples, you can define these 2 groups using -s 's1
                              s2 s3' -s 's4 s5 s6 s7' (where sample name 's1',
                              's2', and 's3' are the wildtype samples and 's4',
                              's5', 's6', and 's7' are mutant samples.  (All
                              sample names must match the sample names in the VCF
                              column headers row.)  The differences in variant
                              states between these groups of samples will be used
                              to sort the variants/rows of the VCF file.  See
                              --extended --help for a description of how degree of
                              difference is calculated.
  
                              ^ If no sample groups are supplied, a default pair
                              of samples that are the most different on any
                              particular row will be chosen.
                              * If only one group is defined, the second group is
                              assumed to be the remainder of the samples.
                              ** If no groups are defined, groups are dynamically
                              determined for each variant/row.  See --help
                              --extended for details.
  
    -d <int,...>...           [all] Each sample group defined by -s is accompanied
    --min-group-size          by a (minimum) number of samples in that group with
      <int,...>...            which to compute the maximum difference against its
                              partner group.  Each instance of -s should have a -d
                              value supplied.  The order of the -d values should
                              correspond to the order of the -s sample groups they
                              apply to.  The default for each group is the group
                              size, but a smaller number can bespecified.  The
                              purpose is best shown by example.  If you have 5
                              mutant samples and 3 replicate wildtype samples, you
                              may want to find variants where 1 or more mutants
                              differ from all 3 wildtype samples, thus -d for the
                              mutant group would be '1' and -d for the wildtype
                              group would be '3'.  In order to produce meaningful
                              results, one group in each pair of groups must get a
                              value that is larger than half the group size.  See
                              --help --extended for details on how this affects
                              variant sorting, sample group growing, and
                              filtering.

    -a <flt>                  [0.6] The maximum difference between average
    --separation-gap <flt>    observation ratios for a given variant state (e.g. a
                              SNP value of "A") between groups defined by -s (and
                              -d) must be at least this value in order to either
                              be retained (see --filter|--nofilter) or grown
                              (--grow|--nogrow).  This option is only used when
                              either --filter or --grow are true.  See --help
                              --extended for more details.
  
    --no-g                    Do not use the genotype call (i.e. the 'GT' value in
    --nogenotype              the FORMAT string) for sorting & filtering rows, or
                              growing sample groups.  See --nogrow and --nofilter.
                              See --help --extended for details.
  
    --no-f                    Do not filter variant rows whose sample group pairs
    --nofilter                do not meet thresholds defined by either `--genotype
                              --min-group-size <int>` or `--nogenotype --min-
                              group-size <int> --separation-gap <float>`.  When
                              --genotype is supplied, the genotype calls in all
                              samples in each minimum group must not have any
                              common genotype calls and meet the minimum size
                              requirement.  When --nogenotype is supplied, the
                              difference in average observation ratios must be
                              greater than or equal to the --separation-gap.  See
                              --help --extended for more details.
  
    --no-w                    If the -d is less than the actual sample group size,
    --nogrow                  by default, the script will keep adding samples to
                              the minimum groups (from its remaining members) as
                              long as (if --nogenotype is supplied) the difference
                              in average observation ratios between the groups is
                              greater than -a or (if --genotype is supplied) the
                              genotype call is the same as current members and
                              different from all partner group genotypes.  Note,
                              this may lower the sort order of a variant/row when
                              --nogenotype is supplied.
  
    ...

## INPUT FORMAT (-i, --vcf-file, --input-file, STDIN)

A VCF file is a plain text, tab-delimited file.  The format is generally described here: http://bit.ly/2sulKcZ and described in detail here: http://bit.ly/2gKP5bN

However, the important parts that this script relies on are:

1. The column header line (in particular - looking for the FORMAT and sample name columns).
2. The colon-delimited codes in the FORMAT column values, specifically (for SNP data produced by freeBayes and Structural Variant data produced by SVTyper) AO (the number of reads supporting the variant), RO (the number of reads supporting the reference), and DP (the number of reads that map at or over the variant position).
3. The colon-delimited values in the sample columns that correspond to the positions defined in the FORMAT column.

The file may otherwise be a standard VCF file containing header lines preceded by '##'.  Empty lines are OK and will be printed regardless of parameters supplied to this script.  Note, the --header and --no-header flags of this script do not refer to the VCF file's header, but rather the run info header of this script.


## OUTPUT FORMAT: (-o)

The output file is the same format as the input VCF files, except sorted differently and possibly filtered.

## OUTPUT FORMAT: (--outfile, -u)

Tab delimited file describing the sorting, filtering, and differing sample groups.  The columns of the file are:

* Chromosome - The chromosome on which the variant is located.
* Position - The position starting from 1 where the variant is located.
* Reference Value - The value the reference has in the variant position.
* Alternate Value(s) - The value(s) observed in the samples in the variant position.
* Best Sample Group Pair Number - The sample group pair's number (numbered from left to right, as they were supplied on the command line) for the pair that resulted in the biggest difference in variant states between the sample groups.
* Best Score - The value the sorting of the file is based on, which is the maximum "Pair Score".
* Best Average Read Depth - The average read depth of the samples in both sample groups of the pair with the best score.
* Pair Number - A colon-delimited list of numbers indicating the pair of sample groups the sort and filtering is based on.
* Pair Score - A colon-delimited list of each pair's maximum difference between sample group 1 & 2.
* Pair Average Read Depth - A colon-delimited list of each pair's average read depth of the samples in both sample groups.
* State(s) Used - The variant states used to calculate the pair score.  The state used can be 0 (the value of the variant at the variant position in the reference), or a number indicating which of the alternate observations produced the best score.  E.g. if in --nogenotype mode and the state '0' was used to compute the scores (indicating the reference state), then the average observation ratio of each group will be close to either N/N (i.e. the same as the reference) and the other group will be close to 0/N (i.e. different from the reference).  In --genotype mode, there are 2 sets of variant state values (i.e. genotype calls) separated by a semicolon.  If there are multiple genotype calls in a sample group, they will be delimited with a "+".
* Sample Group 1 Members - A colon-delimited list of each pair's comma-delimited group 1 sample names that were used to produce the sort value by determining their maximum difference.  E.g. For 2 pairs, the value might be: "s1,s2:s6,s7".
* Sample Group 1 Values - A colon-delimited list of pairs' group 1 comma-delimited values that were used to compute the corresponding "Pair Sort Value".  These may be a series of 0's and 1's indicating whether the sample's genotype is unique to group 1 or they are the observation ratios (in the form "numerator/denominator") of the selected variant state, chosen to differ most between sample groups.
* Sample Group 2 Members - A colon-delimited list of each pair's comma-delimited group 2 sample names that were used to produce the sort value by determining their maximum difference.  E.g. For 2 pairs, the value might be: "s3,s4,s5:s8,s9,s10".
* Sample Group 2 Values - A colon-delimited list of pairs' group 2 comma-delimited values that were used to compute the corresponding "Pair Sort Value".  These may be a series of 0's and 1's indicating whether the sample's genotype is unique to group 1 or they are the observation ratios (in the form "numerator/denominator") of the selected variant state, chosen to differ most between sample groups.

Example:

    #CHROM	POS	ID	REF	ALT	BEST_PAIR	BEST_SEP_SCORE	BEST_AVEDP	PAIR_NUM	PAIR_SEP_SCORE	PAIR_AVEDP	STATE(S)_USED	PAIR1_MEMBERS	PAIR1_SCORE_DATA	PAIR2_MEMBERS	PAIR2_SCORE_DATA
    Chromosome	3413302	.	T	C	1	1	6	1	1	6	1	wtsmpl	2/2	smpl1,smpl2,smpl3	0/9,0/9,0/4
    Chromosome	4904909	.	T	C	1	1	5	1	1	5	1	wtsmpl	2/2	smpl1,smpl2,smpl3	0/5,0/7,0/7
    Chromosome	747041	.	ACCAGCTGGAAAAGGCTGGCGTGAGC	AC	1	1	5	1	1	5	1	wtsmpl	0/7	smpl2,smpl3	5/5,3/3
    Chromosome	3778741	.	T	C	1	1	5	1	1	5	1	wtsmpl	0/7	smpl1	3/3
